launch/launch/actions/clear_launch_configurations.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2021 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module for the ClearLaunchConfigurations action."""
+
+from typing import List
+from typing import Optional
+
+from ..action import Action
+from ..launch_context import LaunchContext
+from ..some_substitutions_type import SomeSubstitutionsType
+
+from ..utilities import perform_substitutions
+
+
+class ClearLaunchConfigurations(Action):
+    """Action that clears the launch configurations from the context."""
+
+    def __init__(
+        self,
+        forwarded_configurations: Optional[List[SomeSubstitutionsType]] = None,
```

Comment:
@djchopp nit: why talk about `forwarded_configurations` here?
```suggestion
        launch_configurations: Optional[List[SomeSubstitutionsType]] = None,
```

Reply:
Maybe more like `launch_configurations_to_be_cleared`. A docstring enty explaining the purpose of this argument would be nice too. I had to read the code to figure out what it was doing.

Reply:
Or, now that I'm looking more closely, I'm not even sure at the moment what the type of these "forwarded_configurations" should be... so definitely need a docstring and maybe even an example in the docstring of how it is meant to be used.

Reply:
Looks more like `launch_configurations_to_not_be_cleared`.

Reply:
@hidmic Good point `launch_configurations` probably makes more sense although @wjwwood suggestion of  `launch_configurations_to_not_be_cleared` is the most descriptive of what it does. 

I modeled its use after `UnsetLaunchConfiguration` where you pass it a list of `LaunchConfiguration` objects as seen in this example:

    def generate_launch_description():
      arg1 = LaunchConfiguration('arg1')

      return LaunchDescription([
        DeclareLaunchArgument(
          'arg1',
          default_value='default_1',
          description='Argument'
        ),
        # Some actions where arg1 is set to 'default_1'
        ClearLaunchConfigurations(launch_configurations_to_not_be_cleared=[arg1])
        # Some more actions where arg1 is not set
      ])

Should I be using a list of strings (keys) instead?

Reply:
Disregard my last comment on using `LaunchConfiguration` objects. I misunderstood how `UnsetLaunchConfiguration` worked. I will change the implementation to match the style of using the key name as a string.

Reply:
Oh, I did not really grok `ClearLaunchConfigurations`. As is stands, `ClearLaunchConfigurations(['arg1'])` means exactly the opposite. I'd avoid that. Following https://github.com/ros2/launch/pull/515#discussion_r658055646, we may not even need this clear-all-but-a-few action.

Reply:
Indeed the more descriptive name for this would be `ClearLaunchConfigurationsExcept`.

See my concern about removing pass through functionality in my [other comment](https://github.com/ros2/launch/pull/515#discussion_r658226793). Specifically not being able to replicate the pass through functionality when using a clear-all instead of a clear-all-but-a-few outside of the group action with special substitution evaluation ordering.

My original use case for this feature was a complex launch system with several layers of depth (includes). Each launch description has many launch configurations. By the time you get to the deepest include, the context launch configurations can be incredibly cluttered. In that deepest include, I want to use mostly default values for the launch configurations. Currently I would have to add an `UnsetLaunchDescription` for each one. This is ok for a couple, but inconvenient past that. The clear-all action works to make sure all defaults are used, but still does not cover wanting to pass just a couple launch configurations through.

If the pass through of just a couple can be accomplished without a clear-all-but-a-few action, then it can definitely be dropped in favor of the clear-all action. I just have not figured out how to do that yet.